accepted/2020/serializer/WriteableDomAndDynamic.md

```diff
@@ -0,0 +1,577 @@
+# Overview and goals
+
+October 30th, 2020.
+
+This document provides options and recommendations for 6.0 to support both a writable DOM and the C# `dynamic` keyword.
+
+Although the writable DOM and `dynamic` features are somewhat separate, considering both during design should result in a single API for both scenarios that is optimized to have minimal state and has intuitive interop between the DOM classes and those returned when using `dynamic`.
+
+The existing `JsonDocument` class represents the DOM support today, however it is read-only. It maintains a single immutable UTF-8 buffer and returns `JsonElement` value types from that buffer on demand, which minimizes the initial `JsonDocument.Parse()` time and heap allocs but also makes it slow to re-obtain the same value (which is common with LINQ) and does not lend itself to being extended to support writability.
+
+This document proposes a new API based on learning and scenarios from:
+- The writable DOM prototype.
+  - During 5.0, there was a [writable DOM effort](https://github.com/dotnet/runtime/issues/30436). Unlike `JsonDocument`, the design centered around writable aspects, usability and LINQ support, and not performance and allocations. It was shelved due to time constraints and outstanding work.
+- The dynamic support code example.
+  - During 5.0, [dynamic support](https://github.com/dotnet/runtime/issues/29690) was not able to be implemented due to schedule constraints. Instead, a [code example](https://github.com/dotnet/runtime/pull/42097) was provided that enables `dynamic` and a writeable DOM that can be used by any 3.0+ runtime in order to unblock the community and gather feedback for a 6.0 feature.
+- Azure prototyping.
+  - The Azure team needs a writable DOM, and supporting `dynamic` is important for some scenarios but not all. A [prototype](https://docs.microsoft.com/en-us/dotnet/api/azure.core.dynamicjson?view=azure-dotnet-preview) has been created. Work is also being coordinated with the C# team to support Intellisense with `dynamic`, although that is not expected to be implemented in time to be used in 6.0.
```

Comment:
> Work is also being coordinated with the C# team to support Intellisense with `dynamic`, although that is not expected to be implemented in time to be used in 6.0.

Issue link in dotnet/roslyn or dotnet/csharplang?

Reply:
@KrzysztofCwalina @MadsTorgersen is there a link based on our discussion from a few days ago?

I see https://github.com/dotnet/roslyn/issues/24226 but not sure if that is the one being used to track the discussion around supporting intellisense through defining "roles" which can be linked to a known schema where the members are known but not necessarily the Type names.